From b7d46252d6d65b0c00b3f859ae5fb827ff8c6683 Mon Sep 17 00:00:00 2001
From: Sandor Yu <R01008@freescale.com>
Date: Thu, 17 Jul 2014 20:18:32 +0800
Subject: [PATCH] ENGR00321817-01 HDMI: Dispaly blank after resume

Issue reproduce steps:
1. Boot up without HDMI cable plugin
2. Insert the HDMI cable.
3. echo mem > /sys/power/state , enter suspend,
4. resume it,
System can resume from suspend but display is blank.
Error log:
mxc_sdc_fb fb.31: Unable to allocate framebuffer memory
detected fb_set_par error, error code: -12

In mxc hdmi driver, if system bootup without hdmi cable plugin,
driver will create a default modelist.

In fbcon driver, array fb_display[] initialized when system bootup
and save current mode pointer that point to default modelist.

When hdmi cable is plugin the modelist will rebuild according edid
data, but the pointer of video mode in fb_display[] is not updated.

When system resume, fbcon will use the invalidate pointer to
configured framebuffer, framebuffer will crash.

Add function fb_new_modelist() after modelist is rebuild to fix the
issue.

Signed-off-by: Sandor Yu <R01008@freescale.com>
(cherry picked from commit 5687cb8dc5099acdb0e3a0542f666326764d558c)
(cherry picked from commit 5451976da30e34db74069d7197748556f9eb5c69)
---
 drivers/video/mxc/mxc_hdmi.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/video/mxc/mxc_hdmi.c b/drivers/video/mxc/mxc_hdmi.c
index 2a816bc..23338b3 100644
--- a/drivers/video/mxc/mxc_hdmi.c
+++ b/drivers/video/mxc/mxc_hdmi.c
@@ -1820,6 +1820,8 @@ static void mxc_hdmi_edid_rebuild_modelist(struct mxc_hdmi *hdmi)
 		}
 	}
 
+	fb_new_modelist(hdmi->fbi);
+
 	console_unlock();
 }
 
@@ -1850,6 +1852,8 @@ static void  mxc_hdmi_default_modelist(struct mxc_hdmi *hdmi)
 			fb_add_videomode(mode, &hdmi->fbi->modelist);
 	}
 
+	fb_new_modelist(hdmi->fbi);
+
 	console_unlock();
 }
 
-- 
1.8.0

